File: python/muse/algorithms/number_theory/Primality.py
import math


def isPrime(n: int) -> bool:
    if n < 2:
        return False
    if (n & 1) == 0 and n != 2:
        return False

    for i in range(3, int(math.sqrt(n)) + 1, 2):
        if n % i == 0:
            return False

    return True


def isComposite(n: int) -> bool:
    return n > 1 and not isPrime(n)

File: rust/muse/src/algorithms/number_theory/primality.rs
pub fn is_prime(n: i64) -> bool {
    if n < 2 {
        return false;
    }
    if (n & 1) == 0 && n != 2 {
        return false;
    }

    let bound = n.isqrt() + 1;
    for i in (3..bound).step_by(2) {
        if n % i == 0 {
            return false;
        }
    }

    true
}

pub fn is_composite(n: i64) -> bool {
    n > 1 && !is_prime(n)
}

#[cfg(test)]
pub(crate) mod tests;
